### PR TITLE
Implement server-side prevalidation session

### PR DESCRIPTION
**Pull request by feihujiang**
Signed-off-by: feihujiang <jiangfeihu@huawei.com>

Implement server-side prevalidation session

### CODE REVIEW

**Comment by feihujiang** [comment]
@vaporos Address part of server-side prevalidation, please review.

**Comment by feihujiang** [comment]
POST /prevalidate
Do server-side prevalidation.

File: extensions/mktplace/mktplace/mktplace_client.py [hunk 1]
```diff
@@ -168,6 +178,9 @@ def _sendtxn(self, update):
         self.LastTransaction = txnid
         txn.apply(self.CurrentState.State)
 
+#         if not self.enable_session:
```
Comment (vaporos) [inline extensions/mktplace/mktplace/mktplace_client.py]: Please remove commented out code.

File: extensions/mktplace/mktplace/mktplace_client.py [hunk 2]
```diff
@@ -189,6 +202,8 @@ def _register(self, update):
         # the name map
         if txnid:
             fqname = self.CurrentState.i2n(txnid)
+#             fqname = update.Name
```
Comment (vaporos) [inline extensions/mktplace/mktplace/mktplace_client.py]: Please remove commented out code.

File: extensions/mktplace/mktplace/mktplace_communication.py [hunk 3]
```diff
@@ -135,12 +137,26 @@ def postmsg(self, msgtype, info):
...
+                    logger.debug('self._cookie %s', self._cookie)
+            else:
+                response = opener.open(request, timeout=10)
 
         except urllib2.HTTPError as err:
             logger.warn('operation failed with response: %s', err.code)
-            raise MessageException(
-                'operation failed with response: {0}'.format(err.code))
+            err_content = err.read()
+            if err_content.find("InvalidTransactionError"):
```
Comment (vaporos) [inline extensions/mktplace/mktplace/mktplace_communication.py]: We should potentially check for a status code of BAD_REQUEST here.

File: extensions/mktplace/mktplace/transactions/market_place.py [hunk 4]
```diff
@@ -495,6 +496,10 @@ def is_valid(self, store):
 
         return True
 
+#     def check_valid(self, store):
```
Comment (vaporos) [inline extensions/mktplace/mktplace/transactions/market_place.py]: Please remove commented out code.

File: validator/setup.py [hunk 5]
```diff
@@ -114,7 +114,7 @@ def version(default):
     url='http://www.intel.com',
     packages=find_packages(),
     install_requires=['sawtooth-core', 'cbor>=0.1.23', 'colorlog',
-                      'twisted', 'PyYAML', 'psutil', 'numpy'],
```
Comment (vaporos) [inline validator/setup.py]: This was perhaps accidentally added to this PR?

**Comment by vaporos** [comment]
The changes to mktclient should (at least eventually) be more extensive, removing it's local state entirely in favor of getting that state from the server.  (And, essentially, removing any direct interaction between mktclient and the mktplace transaction family.)  We don't need to do it all at once, but it is good to keep in mind the goal is to remove that dependency.

**Comment by vaporos** [comment]
Should we explicitly handle sessions with create/delete?  At a minimum, we need some way to clean up sessions from the validator process.

File: extensions/mktplace/mktplace/mktplace_communication.py (hunk 3 above)
Comment (feihujiang) [inline extensions/mktplace/mktplace/mktplace_communication.py]: Done. Others Done too.

**Comment by feihujiang** [comment]
@vaporos I updated this PR, and support creating/deleting sessions, please review. And yes, the goal is removing all of the interaction between mktclient and the mktplace transaction family. But if session is disable, state from the server has no pending transactions now, but some commands need [n2i] (https://github.com/hyperledger/sawtooth-core/blob/master/extensions/mktplace/mktmain/client_cli.py#L600),  so I remain [transaction.apply()](https://github.com/feihujiang/sawtooth-core/blob/897adbdac01aabb7f9924495306ff84bca0f4d8d/extensions/mktplace/mktplace/mktplace_client.py#L185) function. I'll remove it when I have a good idea to handle this case. Or could you give me some advice?

**Comment by danmack** [comment]
Can one of the admins verify this patch?

File: validator/txnserver/web_pages/prevalidation_page.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+import copy
+import logging
+import traceback
+
+from zope.interface import Interface, Attribute, implements
```
Comment (vaporos) [inline validator/txnserver/web_pages/prevalidation_page.py]: We don't want to add a zope dependency here.

**Comment by vaporos** [comment]
+ ./bin/run_lint
************* Module txnserver.web_pages.prevalidation_page
E: 33, 0: Inheriting 'Interface', which is not a class. (inherit-non-class)

File: validator/txnserver/web_pages/prevalidation_page.py (hunk 6 above)
Comment (feihujiang) [inline validator/txnserver/web_pages/prevalidation_page.py]: Why we don't want to add a zope dependency?  And what you suggest?

**Comment by feihujiang** [comment]
@vaporos Could you give some advice if we don't want to add a zope dependency?

**Pull request merged**